src/executor/graph_executor.cc

```diff
@@ -1561,15 +1561,32 @@ static nnvm::Symbol PartitionGraph(const nnvm::Symbol& src,
     } else {
       CHECK(i1 < arg_names.size());
       CHECK_EQ(arg_names[i1], input_names[i]);
-      arg_shapes.push_back(in_args[i1].shape());
-      arg_dtypes.push_back(in_args[i1].dtype());
-      arg_stypes.push_back(in_args[i1].storage_type());
-      in_arg_ctxes[i1] = in_args[i1].ctx();
+      arg_shapes.push_back(in_args->at(i1).shape());
+      arg_dtypes.push_back(in_args->at(i1).dtype());
+      arg_stypes.push_back(in_args->at(i1).storage_type());
+      in_arg_ctxes[i1] = in_args->at(i1).ctx();
       ++i1;
     }
   }
-  return PartitionGraph(src, prop_name, arg_shapes, arg_dtypes, arg_stypes,
-                        default_ctx, ctx_map, in_arg_ctxes, aux_state_ctxes);
+
+  // setup in_args_map
+  std::unordered_map<std::string, NDArray> in_args_map;
+  for (size_t i = 0; i < in_args->size(); ++i) {
+    in_args_map[arg_names[i]] = in_args->at(i);
```

Comment:
@mbrookhart I think MXNet requires inputs with unique names, otherwise, how to distinguish the inputs with same names? According to symbol.bind API:
**args (list of NDArray or dict of str to NDArray) –
Input arguments to the symbol.

If the input type is a list of NDArray, the order should be same as the order of list_arguments().
If the input type is a dict of str to NDArray, then it maps the name of arguments to the corresponding NDArray.
In either case, all the arguments must be provided.** 

So basically, name is the only id for each input.

Reply:
https://github.com/apache/incubator-mxnet/blob/c98b19e2d108a3861d89b475927e8a21a913e540/tests/python/unittest/test_operator.py#L1171

If it's a list, the names don't have to match.

Reply:
In general, MXNet seems to determine node ID's by shared_ptr memory address. Unfortunately, that trick doesn't work with Partition Graph because the pass copies the nodes. :(

Reply:
Yeah, I understand that you can find the order from the code, but it's not guaranteed, and you shouldn't make any assumption based on that, as it may change in future because the order isn't a part of list_arguments() spec, at least for now.

I guess we should answer this question first, which is, should we support inputs with same name? If the answer is shouldn't, then we need to rename those inputs with same name in unit test and add corresponding check and document  to disallow user using like that.

If the answer is we should, then we need to define a clear way for distinguishing inputs apart from name on API level, instead of the undocumented DFS order.

According to the current API design, I guess inputs with same name shouldn't be supported, as the order of list_arguments() is unique only if inputs has unique name. Adding DFS order to list_arguments() spec isn't user friendly, as it's hard for end user to find out the final DFS order from a complex topology, as single op(eg. rand_zipfian) from end-user level may be consist of many small ops in final computing graph.

Reply:
I thought graph partitioning preserves the order of inputs. It now doesn't?

Reply:
@zheng-da Now graph partitioning may change the order of inputs. This is a basic requirement from graph optimization purpose. For example, if we found there's a redundant op that doesn't contribute to graph output, shouldn't we remove it? If the answer is yes, then the input list will be changed.
Besides, current API design provides bad support for inputs with same name(if you treat the undocumented behavior is a kind of support). We should fix this anyway.

Reply:
:) A bug someone is using isn't a bug, it's a feature. I don't know who wrote test_maximum_minimum, but it feels like they were trying to test the duplicate name case?